File: community/cypher/expression-evaluator/src/test/java/org/neo4j/cypher/internal/evaluator/SimpleExpressionEvaluatorTest.java
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.neo4j.cypher.internal.evaluator;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.neo4j.cypher.internal.evaluator.Evaluator.expressionEvaluator;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;
import org.neo4j.cypher.internal.CypherVersion;

class SimpleExpressionEvaluatorTest {

    private final ExpressionEvaluator evaluator = expressionEvaluator(CypherVersion.Default);

    @ParameterizedTest
    @EnumSource(CypherVersion.class)
    void shouldConvertToSpecificType(CypherVersion version) throws EvaluationException {
        assertThat(evaluate(version, "[1, 2, 3]", List.class)).isEqualTo(List.of(1L, 2L, 3L));
    }

    @ParameterizedTest
    @EnumSource(CypherVersion.class)
    void shouldConvertToObject(CypherVersion version) throws EvaluationException {
        assertThat(evaluate(version, "{prop: 42}", Object.class)).isEqualTo(Map.of("prop", 42L));
    }

    @ParameterizedTest
    @EnumSource(CypherVersion.class)
    void shouldThrowIfWrongType(CypherVersion version) {
        // Expect
        assertThatThrownBy(() -> evaluator.evaluate("{prop: 42}", List.class))
                .isExactlyInstanceOf(EvaluationException.class);
    }

    private Object evaluate(CypherVersion version, String expression, Class<?> cls) throws EvaluationException {
        return expressionEvaluator(version).evaluate(expression, cls);
    }
}

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/planning/simpleExpressionEvaluatorTest.scala
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.neo4j.cypher.internal.planning

import org.neo4j.cypher.internal.ast.AstConstructionTestSupport
import org.neo4j.cypher.internal.compiler.planner.logical.simpleExpressionEvaluator
import org.neo4j.cypher.internal.expressions.FunctionInvocation
import org.neo4j.cypher.internal.expressions.FunctionName
import org.neo4j.cypher.internal.util.symbols.CTInteger
import org.neo4j.cypher.internal.util.test_helpers.CypherFunSuite

class simpleExpressionEvaluatorTest extends CypherFunSuite with AstConstructionTestSupport {

  private val randInvocation: FunctionInvocation =
    FunctionInvocation(FunctionName("ranD")(pos), distinct = false, IndexedSeq.empty)(pos)

  test("isNonDeterministic should not care about capitalization") {
    val evaluator = simpleExpressionEvaluator
    evaluator.isDeterministic(randInvocation) shouldBe false
  }

  test("evaluateLongIfStable on IntegerLiteral") {
    simpleExpressionEvaluator.evaluateLongIfStable(literalInt(5)) should be(Some(5))
  }

  test("evaluateLongIfStable on non-deterministic expression") {
    simpleExpressionEvaluator.evaluateLongIfStable(randInvocation) should be(None)
  }

  test("evaluateLongIfStable on expression with parameters") {
    simpleExpressionEvaluator.evaluateLongIfStable(add(literalInt(5), parameter("p", CTInteger))) should be(None)
  }

  test("evaluateLongIfStable on int expression") {
    simpleExpressionEvaluator.evaluateLongIfStable(add(literalInt(5), literalInt(6))) should be(Some(11))
  }

  test("evaluateLongIfStable on string expression") {
    simpleExpressionEvaluator.evaluateLongIfStable(literalString("foo")) should be(None)
  }
}
